### ISSUE TITLE
Cognito User Pool Triggers Disappear On Release

### ISSUE THREAD

**Issue by bowersz** [serverless/serverless#7405]
# Bug Report

## Description

Deployment of my Cognito User Pool that is configured in my yml file is making me nervous...  It seems randomly after a successful deployment, my User Pool Triggers that are also set/configured in my serverless file will disappear. Then, even if I re-deploy, the triggers do not get reset. It's not until I make a change to one of my lambda triggers, or another part of my user pool, that the triggers will be re-set. I'm using the PostConfirmation, PostAuthentication, and CustomMessage triggers. 

What did you do?
I made a change to one of my user pool triggers and then deployed. 

What happened?
After successfully deploying the serverless, I navigate to the AWS Cognito User Pool dashboard and see that the triggers previously set for the user pool have been removed. Then, I re-deploy and the triggers do not get set. I have to make a change to the trigger configuration, or another part of the user pool to reset the triggers.

1. What should've happened?
The trigger I modified should have been set, and the other triggers should not have been modified.

1. What's the content of your `serverless.yml` file?

```
   autoAddUserToDatabase:
    handler: handler.autoAddUserToDatabase
    events:
      - cognitoUserPool:
          pool: ${self:custom.usersUserPoolName}
          trigger: PostConfirmation
          existing: true
  customizeCognitoEmails:
    handler: handler.customizeCognitoEmails
    events:
      - cognitoUserPool:
          pool: ${self:custom.usersUserPoolName}
          trigger: CustomMessage
          existing: true
  autoAddLoginTimeToUser:
    handler: handler.autoAddLoginTimeToUser
    events:
      - cognitoUserPool:
          pool: ${self:custom.usersUserPoolName}
          trigger: PostAuthentication
          existing: true

```

**Comment by medikoo** [comment]
@bowersz thanks for report.

What version of Framework are you using? Can you ensure it's the latest one (?) Also if you rely on any plugins try to reproduce that with none of them involved (it might be plugins influencing the issue).

If you observe this issue on latest framework version with no plugins involved. Please post full content of `serlveress.yml` you're using, with sensitive parts masked out (best if it's as minimal as possible, just to reproduce this issue) and post full output of `sls deploy` when run with `SLS_DEBUG=*`

**Comment by andre-castro-garcia** [comment]
@bowersz @medikoo 
I'm having the same issue in my environment, and I've two files (one for creating the Cognito User Pool and the other for the triggers, which I deploy using serverless deploy).

<details>
<summary><code>resource.yml</code></summary>

```yaml
Resources:
  CognitoSNSRole:
    Type: AWS::IAM::Role
    Properties:
      AssumeRolePolicyDocument:
        Version: '2012-10-17'
        Statement:
          - Effect: Allow
            Principal:
              Service:
                - cognito-idp.amazonaws.com
            Action:
              - sts:AssumeRole
      Policies:
        - PolicyName: !Sub ${StageParam}-cognito-sns-policy
          PolicyDocument:
            Version: '2012-10-17'
            Statement:
              - Effect: 'Allow'
                Action: 'sns:publish'
                Resource: '*'
  CognitoUserPool:
    Type: AWS::Cognito::UserPool
    Properties:
      AccountRecoverySetting:
        RecoveryMechanisms:
          - Name: verified_email
            Priority: 1
          - Name: verified_phone_number
            Priority: 2
      MfaConfiguration: OPTIONAL
      UserPoolName: !Sub ${StageParam}-cognito-general-user-pool
      EnabledMfas:
        - SOFTWARE_TOKEN_MFA
        - SMS_MFA
      SmsConfiguration:
        ExternalId: !Sub ${StageParam}-external
        SnsCallerArn: !GetAtt CognitoSNSRole.Arn
      AutoVerifiedAttributes:
        - email
      UsernameAttributes:
        - email
      Schema:
        - Name: name
          AttributeDataType: String
          Mutable: true
          Required: true
        - Name: family_name
          AttributeDataType: String
          Mutable: true
          Required: true
  CognitoUserPoolDomain:
    Type: AWS::Cognito::UserPoolDomain
    Properties:
      Domain: !Sub ${AWS::AccountId}-${StageParam}-cognito-pool
      UserPoolId:
        Ref: CognitoUserPool
  CognitoUserPoolClient:
    Type: AWS::Cognito::UserPoolClient
    Properties:
      ClientName: !Sub ${StageParam}-api-client
      GenerateSecret: False
      ReadAttributes:
        - name
        - family_name
        - email
        - phone_number
      WriteAttributes:
        - name
        - family_name
        - phone_number
      ExplicitAuthFlows:
        - ALLOW_USER_PASSWORD_AUTH
        - ALLOW_REFRESH_TOKEN_AUTH
      CallbackURLs:
        - http://localhost:3000
      LogoutURLs:
        - http://localhost:3000/logout
      AllowedOAuthFlows:
        - code
      AllowedOAuthFlowsUserPoolClient: True
      AllowedOAuthScopes:
        - openid
        - profile
        - email
        - phone
      SupportedIdentityProviders:
        - COGNITO
      UserPoolId:
        Ref: CognitoUserPool
```

</details>

<details>
<summary><code>serverless.yml</code></summary>

```yaml
functions:
  customMessageHandler:
    name: ${self:StageSuffix}-custom-message-handler
    handler: events/custom-message-handler/index.handler
    description: 'Cognito: Custom Message Handler'
    environment: ${self:custom.environment}
    events:
      - cognitoUserPool:
          pool: ${self:StageSuffix}-cognito-general-user-pool
          trigger: CustomMessage
          existing: True
```

</details>

I'm not sure if it is a bug or a feature request, or some mistake from AWS Cloudformation engine, let me try to explain for you guys what's happening.

1. I deployed the Cognito User Pool using the resource.yml file above (works pretty fine);
2. I deployed the triggers using the serverless.yaml file above (continue works fine);
3. I sent a request to AWS API "aws cognito-idp describe-user-pool <>" and it returns me a Cognito JSON with LambaConfig node with all my triggers;
4. I add a node VerificationMessageTemplate to resource.yaml and upload again (seems to be work fine);
5. I sent a new request to AWS API "aws cognito-idp describe-user-pool <>" and it returns me a Cognito JSON with LambaConfig empty (CloudFormation removed all my triggers);
6. I run "serverless deploy" again, and my triggers don't return (damn it!).

Looking for the files generated by "serverless package" command, I realized that the result file uses a custom resource to attach the trigger to an existing Cognito User Pool. I noticed too that the one custom resource inside a generated Cloudformation file never changes in the next packages, so I guess that CloudFormation does not update the resource (because it is never changed).

```json
{
   "CustomMessageHandlerCustomCognitoUserPool1":{
      "Type":"Custom::CognitoUserPool",
      "Version":1,
      "DependsOn":[
         "CustomMessageHandlerLambdaFunction",
         "CustomDashresourceDashexistingDashcupLambdaFunction"
      ],
      "Properties":{
         "ServiceToken":{
            "Fn::GetAtt":[
               "CustomDashresourceDashexistingDashcupLambdaFunction",
               "Arn"
            ]
         },
         "FunctionName":"dev-custom-message-handler",
         "UserPoolName":"dev-cognito-general-user-pool",
         "UserPoolConfigs":[
            {
               "Trigger":"CustomMessage"
            }
         ]
      }
   }
}
```

In the application inside a lambda, you can notice the same behaviour, look that the function updates, but the custom resource not, this is the reason why my triggers don't come back.

![Screenshot from 2020-05-21 18-52-18](https://user-images.githubusercontent.com/35741698/82610375-9f66c900-9b94-11ea-9ba6-62192979fdc0.jpg)

I was thinking to add some dummy key in the CloudFormation template generated by serverless to force an update. Something like this:
https://github.com/andre-castro-garcia/serverless/blob/0c6c019f4e856919e6aacb4fb4284265fd116190/lib/plugins/aws/package/compile/events/cognitoUserPool/index.js#L179

**Comment by jason-fischer-flashparking** [comment]
This is currently happening for me too using the existing pool option:

```
cognitoPostConfirmation:
  handler: cognito-handler.postConfirmation
  description: Cognito Post Confirmation Trigger
  events:
    - cognitoUserPool:
        pool: my-pool
        trigger: PostConfirmation
        existing: true
```

**"serverless": "1.74.0",
"serverless-bundle": "1.8.0"**

I have disabled all plugins, except serverless-bundle, and get this same behavior.  If the following happens:

- Separate Cognito stack updates UserPool
- or, I manually change the UserPool trigger (setting it to none)

For the last several months, doing either of these and then re-deploying the stack would _always_ put the triggers back, as expected.

Yesterday this seems to have broken on 1.67.0. I updated to 1.74.0, but this has not resolved the issue. The SLS deploy says nothing failed but checking the pool shows it still set incorrectly.  The function is there and available to be picked manually. But the Custom resource stack (that calls the AWS API) doesn't appear to work beyond its first (initial) deploy.

**Comment by medikoo** [comment]
@jason-fischer-flashparking to ensure we coin issue correctly we should keep plugins away.

Anyway this looks as same issue as described here https://github.com/serverless/serverless/issues/7510 where we were trying to discuss problem in more detail. Let's move our discussion there

**Comment by medikoo** [comment]
Duplicate of #7510

**Comment by mattnedrich** [comment]
Does this have any resolution?
